src/semaphore.c

```diff
@@ -716,6 +771,11 @@ _dispatch_thread_semaphore_create(void)
 	int ret = sem_init(s4, 0, 0);
 	DISPATCH_SEMAPHORE_VERIFY_RET(ret);
 	return (_dispatch_thread_semaphore_t)s4;
+#elif USE_FUTEX_SEM
+	dispatch_futex_t *futex = malloc(sizeof(*futex));
```

Comment:
this will create a lot of #if-then-else code

Look at ..
DISPATCH_ALWAYS_INLINE
static inline _dispatch_thread_semaphore_t
_dispatch_get_thread_semaphore(void)
{
    _dispatch_thread_semaphore_t sema = (_dispatch_thread_semaphore_t)
            _dispatch_thread_getspecific(dispatch_sema4_key);
    if (slowpath(!sema)) {
        return _dispatch_thread_semaphore_create();
    }
    _dispatch_thread_setspecific(dispatch_sema4_key, NULL);
    return sema;
}

that would have to be ifdef.
Only other thing I can think off is to use as you say above and return the address
of the thread local variable in the allocation ? Is that's what you meant?

you prefer there to separate out the handling of futex?
I went for cleanness here.